### PR TITLE
Multi_channel: allow more than one instance per program with different configurations

### PR DESCRIPTION
**Pull request by edwintorok**
Multi_channel had shared global state in the form of `dls_new_key` which caused ids assigned by one multichannel to be used by another (possibly smaller channel), resulting in out-of-bounds array indexing.
One possible way to fix this is to remove the global key, and use a per-channel key.

The key here also has some mutexes and conditional variables that you probably don't want shared across different multi-channels.

Draft PR, because there must've been a reason why this was a shared global variable to begin with.

Appears to fix https://github.com/ocaml-multicore/domainslib/issues/43 on my machine

### CODE REVIEW

File: lib/multi_channel.ml [hunk 1]
```diff
@@ -76,13 +78,16 @@ let init_domain_state mchan dls_state =
   [@@inline never]
 
 let get_local_state mchan =
-  let dls_state = Domain.DLS.get dls_key in
-  if dls_state.id >= 0 then dls_state
+  let dls_state = Domain.DLS.get mchan.dls_key in
+  if dls_state.id >= 0 then begin
+    assert (dls_state.id < Array.length mchan.channels);
```
Comment (ctk21) [inline lib/multi_channel.ml:83]: The assert is going to execute every time which is sad, but I guess it can stay if you are strongly in favour.

**Comment by kayceesrk** [comment]
@Sudha247 Perhaps this is relevant to the `Lwt_domain` work.

**Comment by kayceesrk** [comment]
Hi @edwintorok, Jan has created a PR that fixes the conflicts to your branch https://github.com/ocaml-multicore/domainslib/issues/58#issuecomment-998251698. Would you prefer if Jan sends a PR to your `unique-key` branch, which you can merge, after which this PR can be merged?

**Comment by edwintorok** [comment]
Thanks a lot @jmid, I've updated this PR to include your changes.
I've also fixed the test/backtrace.ml unit test which was failing locally (unrelated to this PR, but I like all unit tests to pass).

**Comment by kayceesrk** [comment]
LGTM. Merging.

**Pull request merged**